lib/image_processing/vips.rb

```diff
@@ -0,0 +1,229 @@
+require "vips"
+require_relative "vips/color"
+require_relative "vips/gravity"
+require "tempfile"
+
+module ImageProcessing
+  module Vips
+    module_function
+
+    def self.nondestructive_alias(name, original)
+      define_method(name) do |image, *args, **options, &block|
+        keyword_arguments = options.merge(destructive: false)
+        send(original, image, *args, **keyword_arguments, &block)
+      end
+      module_function name
+    end
```

Comment:
That's an interesting take on the destructive & non-destructive implementation (the `ImageProcessing::MiniMagick` one definitely isn't ideal). But while reading it, I found it a bit strange that a bang method can accept a `:destructive` option to create a new image instead.

Then I came to another idea of solving this problem. Define a non-bang method which accepts a `:destination`. If `:destination` is passed in, then the processing file will be written to that path. If it's not passed in, processed file is written to a new `Tempfile`.  Then we define the bang method that calls the non-bang method and passes the source file path as the `:destination`.

The bonus of that design is that users can then specify a custom destination path, instead of having `ImageProcessing` force a random `Tempfile` path. That way we can also avoid metaprogramming. For example, this is how `convert` could look like:

```rb
def convert(image, format, destination: nil, &block)
  vips_image = ::Vips::Image.new_from_file image.path
  destination_image = destination || _tempfile(".#{format}")
  vips_image.write_to_file(destination_image.path)
  destination_image
end

def convert!(image, format, &block)
  destination_path = Pathname(image.path).sub_ext(".#{format}").to_s
  convert(image, format, destination: File.new(destination_path), &block).tap do
    image.close
    File.delete(image.path)
  end
end
```

And this is how other methods could look like:

```rb
def auto_orient(image, destination: nil, &block)
  with_ruby_vips(image, destination) do |vips_image|
    vips_image.autorot
  end
end

def auto_orient!(image, &block)
  auto_orient(image, destination: image, &block)
end
```

The `with_ruby_vips` method could look something like this:

```rb
def with_ruby_vips(image, destination)
  vips_image = ::Vips::Image.new_from_file image.path
  vips_image = yield(vips_image)
  destination ||= _tempfile
  vips_image.write_to_file(destination.path)
  destination
end
```